skiko/src/awtMain/objectiveC/macos/DisplayLinkThrottler.mm

```diff
@@ -33,19 +34,42 @@ - (instancetype)init {
         _displayLinkScreen = nil;
         _displayLink = nil;
         _vsyncConditionLock = [[NSConditionLock alloc] initWithCondition: 1];
+        _isSleeping = NO;
+
+        NSNotificationCenter *notificationCenter = [[NSWorkspace sharedWorkspace] notificationCenter];
+
+        [notificationCenter addObserver:self
+                            selector:@selector(systemWillSleep:)
+                            name:NSWorkspaceWillSleepNotification
+                            object:nil];
+
+        [notificationCenter addObserver:self
+                            selector:@selector(systemDidWake:)
+                            name:NSWorkspaceDidWakeNotification
+                            object:nil];
     }
 
     return self;
 }
 
+- (void)systemWillSleep:(NSNotification *)notification {
+    _isSleeping = YES;
+    [self invalidateDisplayLink];
+    [self onVSync];
+}
+
+- (void)systemDidWake:(NSNotification *)notification {
```

Comment:
When the system wakes up, we need to notify the window to redraw the content. If system calls the java function `paint()`, it will redraw it. But can we guarantee that? If not, we should send a signal to AWT here.

Reply:
It seems the fix isn't about disabling redrawing, but about disabling vsync. In this case, no need to notify. But will we have unlimited FPS, and should stop drawing in case of sleeping?

Reply:
The issue is about deadlock on the conditional variable encpsulating vsync state. I can't reproduce that so my speculation is to destroy this mechanism temporarily while the device is sleeping.
I have an idea  that have some really funny race-condition, where the logical screen for the window changes mid-flight due to some OS-level reconfigurations (flickering when connecting a display/waking up), so we wait on vsync from the screen which no longer exists/and AppKit didn't gracefully changed the screen to the existing one, but I have no idea currently how to fix _that_ so I'd better try this solution and see if it fixes the issue.

Reply:
Potentially, it can indeed cause discarded drawables (because of redundant redraws, the initial issue that vsync mechanism fixed) if multiple were presented during a single hardware vsync after the the sleep notification, but that's an edge case that I bet will not affect the users

Reply:
vsync also prevents unlimited FPS, which can load GPU/CPU. Do this code run for some time after sleeping?